hypertrace-trace-enricher/trace-reader/src/main/java/org/hypertrace/trace/reader/attributes/DefaultValueResolver.java

```diff
@@ -72,108 +73,106 @@ private Single<LiteralValue> resolveDefinition(
-            .defaultIfEmpty(LiteralValue.getDefaultInstance());
+        return matchingValueSource.flatMap(
+            valueSource -> valueSource.getAttribute(path, attributeKind));
       case METRIC:
-        return matchingValueSource
-            .mapOptional(valueSource -> valueSource.getMetric(path, attributeKind))
-            .defaultIfEmpty(LiteralValue.getDefaultInstance());
+        return matchingValueSource.flatMap(
+            valueSource -> valueSource.getMetric(path, attributeKind));
       case UNRECOGNIZED:
       case TYPE_UNDEFINED:
       default:
-        return this.buildAndLogErrorLazily("Unrecognized projection type");
+        return Optional.empty();
     }
   }
 
-  private Single<LiteralValue> resolveProjection(ValueSource valueSource, Projection projection) {
+  private Optional<LiteralValue> resolveProjection(ValueSource valueSource, Projection projection) {
     switch (projection.getValueCase()) {
       case ATTRIBUTE_ID:
-        return valueSource
-            .executionContext()
-            .wrapSingle(() -> this.attributeClient.get(projection.getAttributeId()))
+        return this.attributeClient
+            .getById(valueSource.requestContext(), projection.getAttributeId())
             .flatMap(attributeMetadata -> this.resolve(valueSource, attributeMetadata));
       case LITERAL:
-        return Single.just(projection.getLiteral());
+        return Optional.of(projection.getLiteral());
       case EXPRESSION:
         return this.resolveExpression(valueSource, projection.getExpression());
       case VALUE_NOT_SET:
       default:
-        return this.buildAndLogErrorLazily("Unrecognized projection type");
+        if (LOGGING_LIMITER.tryAcquire()) {
+          log.error("Unrecognized projection type");
+        }
+        return Optional.empty();
     }
   }
 
-  private Single<LiteralValue> resolveField(
+  private Optional<LiteralValue> resolveField(
       ValueSource valueSource, SourceField sourceField, AttributeKind attributeKind) {
-    return Maybe.fromOptional(valueSource.getSourceField(sourceField, attributeKind))
-        .defaultIfEmpty(LiteralValue.getDefaultInstance());
+    return valueSource.getSourceField(sourceField, attributeKind);
   }
 
-  private Single<LiteralValue> resolveFirstValuePresent(
+  private Optional<LiteralValue> resolveFirstValuePresent(
       ValueSource valueSource,
       AttributeMetadata attributeMetadata,
       AttributeDefinitions definitions) {
 
-    return Observable.fromIterable(definitions.getDefinitionsList())
-        .concatMapMaybe(
-            definition -> this.maybeResolveDefinition(valueSource, attributeMetadata, definition))
-        .first(LiteralValue.getDefaultInstance());
+    return definitions.getDefinitionsList().stream()
+        .map(definition -> this.resolveDefinition(valueSource, attributeMetadata, definition))
+        .filter(Optional::isPresent)
+        .findFirst()
+        .flatMap(Function.identity());
```

Comment:
if I'm reading this correctly you've got a resolution that can return an optional and you're just trying to get the first value present, if it exists. If so, suggest:


```suggestion
        .flatMap(Optional::stream)
        .findFirst();
```